Add option for ignoring blank lines for report line indentation

SamKacer opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/wiki/Contributing.
-->

### Link to issue number: 

closes #13394

### Summary of the issue:

As elaborated in #13394, line indentation reporting can be too noisy in contexts where indentation changes for blank lines aren't meaningful and distract from the actually meaningful indentation changes, such as is the case when reading source code for Python and many other programming languages.

### Description of user facing changes

A new checkbox is added to Document formatting, "Ignore blank lines for line indentation reporting". By default it is off, in which case line indentation reporting behaves as it does currently. If the user ticks the checkbox, then indentation changes are not reported for blank lines.

There is also a new unassigned global command for toggling the setting.

### Description of development approach

- new config property: `ignoreBlankLinesForRLI = boolean(default=false)`
  - short for `ignoreBlankLinesForReportLineIndentation`, which was too long and causing Flake8 issues
- new settings panel checkbox for toggling `ignoreBlankLinesForRLI`
- in `speech.speech.getTextInfoSpeech`, if `ignoreBlankLinesForRLI` is true and the line is blank, the block of code that inserts the indentation announcement and updates the indentation cache is skipped, otherwise it behaves as usual
- user guide updated with brief info on new option
- added unassigned global command for toggling `ignoreBlankLinesForRLI`
- added system test for confirming the correct behavior

### Testing strategy:

I added a new test in `symbolPronunciationTests` that confirms the correct behavior by comparing speech output when the new setting is on/off.

I also tested manually by reading the following text with "Ignore blank lines for line indentation reporting" turned off and on. In either case make sure to have line indentation reporting switched on. To quickly find the new setting, open document formatting settings with NVDA + Shift + D and press Alt + I to get to "line indentation reporting", then press tab to get to "Ignore blank lines for line indentation reporting". Alternatively, you can press alt + B twice.

#### Test text:

```
def foo():
	x = 42

	return x

def bar():
```

With current NVDA or with this change and "Ignore blank lines for line indentation reporting" turned off, the text should report 4 indentation changes like:

```
def foo():
tab x = 42
no indent blank
tab return x
no indent blank
def bar():
```

With "Ignore blank lines for line indentation reporting" turned on, it should instead report just 2 indentation changes like:

```
def foo():
tab x = 42
blank
return x

no indent def bar():
```

### Known issues with pull request:

None that I am aware of.

### Change log entries:
New features
- Add new option to document formatting settings, "Ignore blank lines for line indentation reporting" (#13394 )
Changes
Bug fixes
For Developers

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

SamKacer committed:
Implement ignore blank lines for report line indentation

SamKacer committed:
Add setting checkbox for ignoreBlankLinesForReportLineIndentation

SamKacer committed:
Update user doc about ignore blank lines

SamKacer committed:
Add additional info on ignore blank lines in user guide

XLTechie commented:
Please add an accellerator key for this checkbox. I suggest Alt+b.

brunoprietog commented:
Nice! I hope this will be merged into NVDA. Just a question, instead of the indentation read change occurring on the last line, will it be possible to have NVDA say something like this?

```
def foo():
tab x = 42
blank
return x
no indent blank
def bar():
```

That makes a lot more sense to me. Thanks for bringing this to NVDA!

XLTechie commented:
In testing this build, it performs in every way as I would expect.

* Tested with tones and speech (same).
* Tested switching to an editor window while on a line with an indent in that editor window.
* Tested changing to a window while on an indented line, at the same indent level as the window I departed.
* Tested moving to an indented line from a blank line, immediately after switching windows.

While I haven't reviewed the code, the behavior is good IMO, and a useful feature to have.

XLTechie commented:
@brunoprietog How would it know which blanks were safe to ignore, and which blanks to reset on? I agree that your wish would make it a more useful feature, but it seems impossible to implement.

That said, if it were a combobox with options of:

* Speak normally (default)
* Ignore
* Ignore only one line

Then that third option could turn something like:

```
def foo():
	x = 42
	return x


def bar():
	pass
```

Into:

```
def foo():
tab  x = 42
return x
blank
no indent blank
def bar():
tab  pass
```

Which is more Pythonic (consistent with pep-8) for top-level functions anyway, though won't help you with methods.

brunoprietog commented:
It's difficult. What occurs to me is that as the user listens to the lines, NVDA checks what is the indentation level of the next line that is not blank in the direction the user navigated (up or down).

If the user presses the up or down arrow, NVDA should check the indentation level of the next line that is not blank, in the direction in which the user navigated (up or down).
- If the indentation level is lower than the current one, on the next blank line NVDA should immediately report the indentation level change, showing that it is now lower, even if it is on a blank line.
- If the indentation level is higher than the current one, NVDA should notify the user of the indentation change to the higher level when the user reaches that line.

Here is an example:

```python
def foo
	pass



def bar
	pass

def baz


	pass
```

Reading it from top to bottom, NVDA should say this:

```
def foo
tab pass
no indent blank
blank
blank
def bar
tab pass
no indent blank
def baz
blank
blank
tab pass
```

Reading it from the bottom up, NVDA should say this:

```
tab pass
no indent blank
blank
def baz
blank
tab pass
no indent def bar
blank
blank
blank
tab pass
no indent def foo
```

That's more intuitive for me, at least in my case

cary-rowen commented:
Hi @SamKacer 

This is a nice change and I like it.
I believe this will be beneficial to all, Whether this option can be checked by default.

Thanks

XLTechie commented:
NVDA has never tried to be predictive about what navigation the user might do 
next. I understand why you want this, and it even seems reasonable. But that is 
a more controversial change than the original request I think.

I would rather see this go forward as is (or with my one-line-only 
modification), then see the proposal to be predictive as a feature request after 
this feature is established.

All kinds of things open up if we're willing to predict what the user might want 
to know or do next, but I think that will need deeper discussion than is called 
for by this simple change. A proposal like that has potentially concerning 
side-effects, and will need exploration.

leonardder commented:
How ought this to behave when indentation reporting with tones is on?

SamKacer commented:
@XLTechie wrote
> Please add an accelerator key for this checkbox. I suggest Alt+b.
> 
> Also, would an unassigned gesture to toggle it be useful?

Thanks, will do those when I have time. This is my first time adding any new config property, so please let me know about anything I missed. Actually, how do I add the accelerator key? I haven't noticed that when making my changes. Also in which file are the commands for toggling setting properties defined?

to do:

- [ ] add accelerator key for new setting
- [ ] add unassigned toggle command for new setting

XLTechie commented:
> how do I add the accelerator key? I haven't noticed that when making my changes.

Ampersand before the letter in the setting name. So if it was "Announce blanks", 
and you wanted the key to be alt+b, you would write it as "Announce &blanks".

Btw, while you're in gui/settingsDialogs.py, please consider fixing the spelling 
of "panel" on lines 2393 and 2400.

> Also in which file are the commands for toggling setting properties defined?

globalCommands.py

See script_toggleProgressBarOutput for an example.
Also see script_reportLinkDestinationInWindow, for an example of proper typing, 
which isn't done in the other one.

XLTechie commented:
> How ought this to behave when indentation reporting with tones is on?

Based on testing it, it behaves exactly as I expected.

There are no de-indent beeps on the lines that are blank.

SamKacer commented:
In regards to @brunoprietog request. He has already asked this in #13394 and I gave a very long reply there [in this comment](https://github.com/nvaccess/nvda/issues/13394#issuecomment-1059757162). Linking in case someone wants full context.

Thanks for the suggestion, but this would require a significant overhaul of the indentation reporting code. I am not entirely sure if I would even prefer that behavior. I think I might prefer if the rules for indentation reporting are kept simpler without special cases like this. In any case, I'm not sure the amount of effort to implement would be that worth it.

SamKacer commented:
> How ought this to behave when indentation reporting with tones is on?

@leonardder the behavior should be the same whether you are using tones or speech

SamKacer committed:
Add mnemonic for ignore blank lines and fix some spellings

SamKacer committed:
Add unassigned command for toggling ignore blank lines

SamKacer committed:
Revise user guide for ignore blank lines

SamKacer committed:
Add system test for ignore blank lines setting

SamKacer committed:
Remove trailing whitespace

SamKacer committed:
Fix Flake8 issues

SamKacer committed:
Fix error in ignore blank lines system test

SamKacer committed:
Fix system test and Flake8

SamKacer committed:
Shorten config prop name to ignoreBlankLinesForRLI

SamKacer commented:
Ok, this is officially ready for review.

I implemented the 2 features @XLTechie suggested. I added a system test. I fixed all the Flake8 issues. The summary is up to date.

Please let me know if anything else is missing or if anything could be improved. Thanks

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -2118,6 +2119,9 @@ The tone will increase in pitch every space, and for a tab, it will increase in
 -
 
+If you tick the "Ignore blank lines for line indentation reporting" checkbox, then indentation changes won't be reported for blank lines.
+This may be useful when reading a document where blank lines are used to separate indented bloks of text, such as in programming source code.
```
Fix typo

```suggestion
This may be useful when reading a document where blank lines are used to separate indented blocks of text, such as in programming source code.
```

seanbudd commented on tests/system/robot/symbolPronunciationTests.py:
```diff
@@ -536,20 +543,32 @@ def test_symbolInSpeechUI():
 
-def _setSymbolLevel(symbolLevel: SymLevel) -> None:
+def _setConfig(
+		symbolLevel: SymLevel,
```
Now that this is more complex, can this get a sensible default too to match the default in configSpec, `SymLevel.SOME`

```suggestion
		symbolLevel: SymLevel: SymLevel.SOME,
```

SamKacer replied:
I am just wondering if this adds value here. Adding defaults for the 2 new properties was helpful because that way I didn't need to make as many changes. Every place that uses `_doTest` or `_setConfig` already specifies symbol level. Also although the default in `configSpec` for sym level is some, this value isn't actually used in any of these tests. So isn't adding a default for that param unnecessary and if it was to be added, wouldn't NONE or ALL be a more useful default value for these tests?

seanbudd replied:
I think it would be useful to have here in-case future tests do not wish to change the default value for symbol level, and instead wish to test the default symbol level.
Previously the function was `_setSymbolLevel` which was for explicitly changing the symbol from default. Now that it is `_setConfig` and empty call to `_setConfig` should not change the default config.

SamKacer replied:
I think that definitely makes sense. I am just concerned that this could be overengineering, since this is a change to support a testing pattern that isn't being used and might never be.

Consider that if I accepted this code change, the subsequent build would fail. Beyond the syntax being incorrect, SymLevel doesn't actually have a `SOME` member. It just has `NONE` and `ALL`, because that is the only options used in these tests. So if I was to make this change, I would also have to add that member. But since I would be adding it just in case someone in the future wanted to use it, then wouldn't it make sense to also add `MOST` in case someone wanted to use that in the future. And then it would make most sense to fill in the other enums like `IndentationReporting` with their missing options.

Instead wouldn't it be better if whoever would have such a use case in the future to be the one to add these things and make the changes that are necessary? I think that would avoid making changes and additions that might not even end up being used.

seanbudd replied:
I'd rather not accept a misleading function into the code base. I think the change of purpose to `_setConfig` warrants the change. The only additional required code involves adding a `SOME` member.

SamKacer replied:
But isn't it misleading to add the `SOME` member if it isn't actually being used in any test? I am not sure if I understand how `_setConfig` is now misleading. Back when it was `_setSymbolLevel` it didn't have any dfault and it wasn't considered misleading then. If the name of the function is what is misleading, then what if it was changed to something like `_changeConfig` or `_overrideConfig`?

@michaelDCurran @Qchristensen could we please get a third opinion on this thread?

SamKacer replied:
Alternatively, another option could be to return `_setSymbolLevel` to what it was and add a `_setIndentationReporting` function.

michaelDCurran replied:
Since we are dealing with configuration values here, I think they should either all be explicit, or all be optional with defaults.
As you don't wish to repeat indentation values that won't change much, I think the only option then is to make everything optional as @seanbudd  said.
However, I also like the alternative of adding _setIndentationReporting and keeping these functions separate.
So in short, if you are unwilling to make the symbol level value optional, then perhaps go with _setIndentationReporting.
However, as @seanbudd  is the NvDA core developer reviewing this pr, he should have the final say.

SamKacer replied:
@michaelDCurran thanks for your input

@seanbudd would you kindly make the code change you had in mind?

seanbudd replied:
That's been done

seanbudd commented on source/globalCommands.py:
```diff
@@ -622,6 +622,23 @@ def script_toggleReportLineIndentation(self, gesture: inputCore.InputGesture):
+		description=_("Toggles on and off the ignoring of blank lines for line indentation reporting"),
+		category=SCRCAT_DOCUMENTFORMATTING
+	)
+	def script_toggleignoreBlankLinesForReportLineIndentation(self, gesture: inputCore.InputGesture) -> None:
```
Fixup CamelCase for `toggleIgnore`

```suggestion
	def script_toggleIgnoreBlankLinesForReportLineIndentation(self, gesture: inputCore.InputGesture) -> None:
```

seanbudd commented on tests/system/robot/symbolPronunciationTests.py:
```diff
@@ -536,20 +543,32 @@ def test_symbolInSpeechUI():
 		navKey: Move,
 		expectedSpeech: _typing.List[str],
 		reportedAfterLast: EndSpeech,
 		symbolLevel: SymLevel,
```
same here re-sensible defaults

SamKacer commented on source/speech/speech.py:
```diff
@@ -1540,7 +1540,17 @@ def getTextInfoSpeech(  # noqa: C901
+			# either not ignoring blank lines
+			not formatConfig['ignoreBlankLinesForRLI']
+			# or line isn't completely blank
+			or any(isinstance(t, str) and not all(c in LINE_END_CHARS for c in t) for t in textWithFields)
```
Isn't this a case perhaps where having this expression on one line makes things more readable? The preceding line succintly summarizes what that expression computes. I am imagining in most cases when someone is reading this part of the code they just want to understand what is happening at a high level. Spreading out this expression over multiple lines, particularly since it computes a simple and very straightforward thing means a reader has to arrow through more lines of code. When it is on one line, they can skip past it with one key stroke if they don't care about the details (probably 90% of the time) and if they do care about the details then they can move by word.

Alternatively, what if this expression was moved into a helper function where the expression was spread out more as you suggest, but that way this bit of code where it would be called would remain succinct?

seanbudd replied:
a helper function may be better here - but the current line is not parseable for me at all. I had to separate it to newlines understand the change here.

SamKacer committed:
Make expression more readable

seanbudd committed:
Merge remote-tracking branch 'origin/master' into ignoreBlankLinesForReportLineIndentation

seanbudd committed:
fix up tests - add symLevel sensible default

Qchristensen reviewed (approved):
User guide reads well, thanks!

seanbudd committed:
Merge remote-tracking branch 'origin/master' into ignoreBlankLinesForReportLineIndentation

seanbudd committed:
update changes
